### ISSUE TITLE
Fixed module position that is not affected by length of output

### ISSUE THREAD

**Issue by rieje** [jaagr/polybar#1690]
Modules constantly shift when output length changes. For example, when volume changes from 0 to 50 to 100 (from 1 digit to 2 digits to 3 digits), the volume module causes all modules to its left to shift left (assuming the volume module is at modules-right.

I tried adding padding but it doesn't help:

    [module/pulseaudio]
    type = internal/pulseaudio

    use-ui-max = false

    format-volume = <label-volume>
    label-volume = VOL %percentage:3:3%

    label-muted = MUTE %percentage%
    label-muted-foreground = #666

A possible solution is to allow parts of the output to be aligned to the left or right. For example, the output for volume may look like this if the value of the volume is aligned to the right. Here's an [example](https://pastebin.com/dUB2Sztp).

Such a feature is useful because the status bar serves to deliver information to the user as efficiently and seamlessly a possible. Shifting modules are distracting and clutter where the user expects to see certain kinds of information.

**Comment by Lomadriel** [comment]
For the moment there is no such option, but if it doesn't bother you, you can add an extra 0 before the minimum width of the `percentage` token. The string will be 0-padded.

```ini
[module/pulseaudio]
type = internal/pulseaudio

use-ui-max = false

format-volume = <label-volume>
label-volume = VOL %percentage:03:3%

label-muted = MUTE %percentage%
label-muted-foreground = #666
```

It will display the volume like that `VOL 005%. However it could be a good idea to add a feature to customize this padding.

(For the alignment I think this is complicated since module doesn't have size)

**Comment by rieje** [comment]
It does bother me, but I'm glad there's a decent workaround, thanks. 

I assume it's not possible to style the padding part of the label to make the padding the same color as the background of the bar, effectively making it invisible? For example, I tried `label-volume = VOL %{F#a1a1a1}%percentage:03%{F-}:3%` and it doesn't work. That could be another decent fix rather than a implement a specific padding setting.

**Comment by Lomadriel** [comment]
Okay no there is another solution, I was wrong sorry, forget what I said before.

If you write `%percentage:3:3%`, the default padding is a whitespace **_AND_** if you use a Monospace font, each character will have the same size, so your numbers will have the same size than a space, and It will behave like you want.

**Comment by patrick96** [comment]
As @Lomadriel correctly mentioned `%percentage:3:3%` does exactly what you want, if you use monospace fonts.

**Comment by rieje** [comment]
I've tried `%percentage:3:3%` as mentioned when submitting this issue or feature request, along with `%percentage:3%`. Modules still shifting. So I guess it's a bug because it's not working for me? I am using the font Inconsolata to test (Monospace font). `%percentage:03:3%` as mentioned by @Lomadriel was a "workaround" only in the sense that it shows the volume and doesn't move the modules, but is ugly because a volume of 1 is 001.

polybar -vvv:

```
polybar 3.3.0

Features: +alsa +curl +i3 -mpd +network(libnl) +pulseaudio +xkeyboard

X extensions: +randr (+monitors) -render -damage -sync +composite +xkb +xrm +xcursor

Build type: RelWithDebInfo
Compiler: /usr/bin/c++
Compiler flags: -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -Wall -Wextra -Werror -Wno-noexcept-type -pedantic -pedantic-errors
Linker flags: -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now
```

[My config](https://pastebin.com/1vKwjLNS).

**Comment by Lomadriel** [comment]
Could you post the output of `polybar -l info -c [your config path]` please.

**Comment by rieje** [comment]
`polybar -l info -c ~/.config/polybar/config`:

```
Usage: polybar [OPTION]... BAR

  -h, --help               Display this help and exit
  -v, --version            Display build details and exit
  -l, --log=LEVEL          Set the logging verbosity (default: WARNING)
                           LEVEL is one of: error, warning, info, trace
  -q, --quiet              Be quiet (will override -l)
  -c, --config=FILE        Path to the configuration file
  -r, --reload             Reload when the configuration has been modified
  -d, --dump=PARAM         Print value of PARAM in bar section and exit
  -m, --list-monitors      Print list of available monitors and exit
  -w, --print-wmname       Print the generated WM_NAME and exit
  -s, --stdout             Output data to stdout instead of drawing it to the X window
  -p, --png=FILE           Save png snapshot to FILE after running for 3 seconds
```

`fc-match -a | grep -i inconsolata`:

```
Inconsolata-Regular.ttf: "Inconsolata" "Regular"
Inconsolata-Bold.ttf: "Inconsolata" "Bold"


```

**Comment by Lomadriel** [comment]
I tried to reproduce the issue with your config and it works as intended, I launched your config that way:
`polybar -c [your config] vega`

However if you use the `zeta` bar, indeed it won't work, the default font for this bar is not the monospace font, it is `FontAwesome`, if I add `label-volume-font = 2` or swap the fonts, there is no problem. I advise you to swap the first two fonts to have you text font as default. (You may need to do some adjustments especially in your i3 module where you have a padding of 6)

So for me there is no bug, feel free to update the post, if I am wrong.

**Comment by rieje** [comment]
As suggested, changing `label-mode-padding` to 0 solved it for me, thanks.

I suppose the original feature request of fixed module position or length is still standing so you aren't  limited to Monospaced fonts if non-moving modules are desired, but that's less of a priority so issue can remain closed.

**Comment by amplexus** [comment]
I had a similar problem. I was convinced it was a bug til I actually followed the troubleshooting steps in https://github.com/polybar/polybar/wiki/Fonts#debugging-font-issues and discovered after having a stable config for many months that polybar was never using the mono font I nominated because I didn't specify it properly and so polybar ended up using a fallback font that wasn't mono. Once I specified the font correctly the shifting disappeared!
